src/client.rs

```diff
@@ -6,33 +6,36 @@ use std::sync::Arc;
 use anyhow::{bail, format_err, Context, Error};
 use base64::prelude::BASE64_STANDARD_NO_PAD;
 use base64::Engine;
+use bytes::Bytes;
 use ed25519_dalek::SigningKey;
 use http::header::{AUTHORIZATION, CONTENT_TYPE};
 use http::request::{Builder, Parts};
 use http::{HeaderValue, Uri};
-use hyper::body::{to_bytes, HttpBody};
-use hyper::client::connect::Connect;
-use hyper::{Body, Client, Request, Response};
+use http_body_util::{BodyExt, Full};
+use hyper::{Request, Response};
+use hyper_util::client::legacy::connect::Connect;
+use hyper_util::client::legacy::Client;
+use hyper_util::rt::TokioExecutor;
 use serde::Serialize;
 use serde_json::value::RawValue;
 use signed_json::signed::Wrap;
 use signed_json::{Canonical, CanonicalWrapper, Signed};
 
 use crate::server_resolver::{handle_delegated_server, MatrixConnector};
 
-/// A [`hyper::Client`] that routes `matrix://` (Synapse <1.87.0rc1 (2023-06-27)) and
-/// `matrix-federation://` (Synapse >=1.87.0rc1 (2023-06-27)) URIs correctly, but does
-/// not sign the requests.
+/// A [`hyper_util::client::legacy::Client`] that routes `matrix://` (Synapse <1.87.0rc1
+/// (2023-06-27)) and `matrix-federation://` (Synapse >=1.87.0rc1 (2023-06-27)) URIs
+/// correctly, but does not sign the requests.
 ///
 /// Either use [`SigningFederationClient`] if you want requests to be automatically
 /// signed, or [`sign_and_build_json_request`] to sign the requests.
 #[derive(Debug, Clone)]
 pub struct FederationClient {
-    pub client: hyper::Client<MatrixConnector>,
+    pub client: Client<MatrixConnector, Full<Bytes>>,
```

Comment:
`Full<Bytes>` seems like the meta looking at the [`hyper-util` examples](https://github.com/hyperium/hyper-util/blob/46826ea75836852fac53ff075a12cba7e290946e/src/client/legacy/client.rs#L190-L214) and  [`hyper` examples](https://github.com/hyperium/hyper/tree/621d8e4d7788bfd2d62d15d40a73efae7f9a0bf0/examples).

Reply:
The `B` here is for the body of the *request* right? My only concern is that generally you don't want to use `Full<_>` style APIs when receiving data from remotes, as then stuff gets buffered up in memory and it becomes easy to DoS. That's not a problem if its just used by the calling code

Reply:
*Re-reading, it looks like this could use a reply.*

From the `hyper` code itself, it's really unclear and they don't have any doc comments for it but in all of our usage, it's only `Request<Full<Bytes>>` and `Response<hyper::body::Incoming>` which matches your hopes.